### PR TITLE
Close #19: Add `short-space-opt` as optional feature

### LINKED ISSUE #19: Optional space between key and value
Hi, we are using this library for the [Saltwater C Compiler](https://github.com/jyn514/saltwater) and I wanted to ask if it would be possible to make the space between key and value optional for short form arguments (e.g. -I/usr/local/include). These are commonly used and it would increase compatibility with other compilers. I am willing to implement this as an optional feature like `eq-separator`; I just want to check if there is a reason it has not been implemented yet.

### PR DESCRIPTION
**Pull request by hdamron17**
Close #19. This will add `short-space-opt` as an optional feature which allows the space between key and parameter to be left out (more details are in the readme). Let me know if there are any additional tests I should include or any changes you would like. Also, please let me know if there is a way to clean up the `cfg` conditions for using `eq-separator` and `short-space-opt` together.

### CODE REVIEW

File: tests/tests.rs [hunk 1]
```diff
@@ -189,14 +189,54 @@ fn eq_option_err_07() {
...
+fn space_option_01() {
+    let mut args = Arguments::from_vec(to_vec(&["-w10"]));
+    let value: Option<u32> = args.opt_value_from_str("-w").unwrap();
+    assert_eq!(value.unwrap(), 10);
+}
+
+#[cfg(feature = "short-space-opt")]
+#[test]
+fn space_option_02() {
+    let mut args = Arguments::from_vec(to_vec(&["-w--width"]));
```
Comment (RazrFalcon) [inline tests/tests.rs:211]: This is very error prone... But I guess we can't do anything about it.

File: tests/tests.rs [hunk 2]
```diff
@@ -189,14 +189,54 @@ fn eq_option_err_07() {
...
+    assert_eq!(value.unwrap(), "10");
+}
+
+#[cfg(feature = "short-space-opt")]
+#[test]
+fn space_option_err_01() {
+    let mut args = Arguments::from_vec(to_vec(&["--width10"]));
+    let value: Option<String> = args.opt_value_from_str("--width").unwrap();
+    assert_eq!(value, None);
+}
```
Comment (RazrFalcon) [inline tests/tests.rs:238]: What about quoted values? Like `-a'b'`?

File: src/lib.rs [hunk 3]
```diff
@@ -743,7 +750,35 @@ fn starts_with_plus_eq(text: &OsStr, prefix: &str) -> bool {
...
+        if s.get(0..prefix.len()) == Some(prefix) {
+            return true;
+        }
+    }
+
+    false
+}
+
+#[cfg(all(feature = "eq-separator", feature = "short-space-opt"))]
+fn index_predicate(text: &OsStr, prefix: &str) -> bool {
```
Comment (RazrFalcon) [inline src/lib.rs:769]: All three of this methods can be inlined.

File: src/lib.rs [hunk 4]
```diff
@@ -312,18 +312,25 @@ impl Arguments {
...
 
             let value = &self.0[idx];
 
             // Only UTF-8 strings are supported in this method.
             let value = value.to_str().ok_or_else(|| Error::NonUtf8Argument)?;
 
             let mut value_range = key.len()..value.len();
-            if value.as_bytes().get(value_range.start) == Some(&b'=') {
+
+            #[cfg(feature = "eq-separator")]
```
Comment (RazrFalcon) [inline src/lib.rs:324]: All the `eq-separator` code can be move into a it's own block, e.g.:

```rust
#[cfg(feature = "eq-separator")]
{
    if value.as_bytes().get(value_range.start) == Some(&b'=') {
        value_range.start += 1;
    } else {
        // Key must be followed by `=`.
        return Err(Error::OptionWithoutAValue(key));
    }
}
```

File: src/lib.rs (hunk 3 above)
Comment (hdamron17) [reply src/lib.rs:768]: That's what I thought, but I wasn't sure if inlining worked in Rust the same as it does in C.

File: tests/tests.rs (hunk 2 above)
Comment (hdamron17) [reply tests/tests.rs:262]: New tests added.

File: src/lib.rs (hunk 4 above)
Comment (hdamron17) [reply src/lib.rs:324]: Thanks! I tried this without making it a block and the compiler complained that this wasn't in stable rust yet, but I guess it's because `if` statements can also act as expressions if they're not in their own block.

File: tests/tests.rs (hunk 1 above)
Comment (hdamron17) [reply tests/tests.rs:211]: I agree, but that's on the end user to not mess up.

File: tests/tests.rs [hunk 5]
```diff
@@ -189,14 +189,86 @@ fn eq_option_err_07() {
...
+    assert_eq!(value.unwrap(), 10);
+}
+
+#[cfg(all(feature = "short-space-opt", not(feature = "eq-separator")))]
+#[test]
+fn space_not_eq_option_01() {
+    let mut args = Arguments::from_vec(to_vec(&["-w=10"]));
+    let value: Option<String> = args.opt_value_from_str("-w").unwrap();
+    assert_eq!(value.unwrap(), "=10");
+}
```
Comment (RazrFalcon) [inline tests/tests.rs:238]: This is very bad. Cargo features are additive, so we cannot change the behavior that much and must assume that a user have enabled all the flags.

I think we should return an error, when `eq-separator` is disabled, but `=` is present. Otherwise, it can silently return `10` instead of `=10`.

Comment (RazrFalcon) [reply tests/tests.rs:246]: Looks like I was not clear. You should have used the `OptionWithoutAValue` error. Adding a new error is a breaking change and I don't want that.

Comment (hdamron17) [reply tests/tests.rs:246]: I added an error for the edge case when `short-space-opt` is enabled but `eq-separator` is not and the user gives `-w=10`. If the user really wants to give '-10' then it can be quoted. The tests reflect this behavior.

Comment (hdamron17) [reply tests/tests.rs:247]: My bad, I used `OptionWithoutAValue` in the latest commit.

**Comment by RazrFalcon** [comment]
Thanks! Everything looks good now.

**Comment by hdamron17** [comment]
Awesome! Thanks so much for your help!
